glean-core/src/histogram/functional.rs

```diff
@@ -0,0 +1,136 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+use std::collections::HashMap;
+
+use serde::{Deserialize, Serialize};
+
+use super::{Bucketing, Histogram};
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct Functional {
+    exponent: f64,
+}
+
+impl Functional {
+    fn new(log_base: f64, buckets_per_magnitutde: f64) -> Functional {
+        let exponent = log_base.powf(1.0 / buckets_per_magnitutde);
+
+        Functional { exponent }
+    }
+
+    fn sample_to_bucket_index(&self, sample: u64) -> u64 {
+        ((sample + 1) as f64).log(self.exponent) as u64
+    }
+
+    fn bucket_index_to_bucket_minimum(&self, index: u64) -> u64 {
+        self.exponent.powf(index as f64) as u64
+    }
+}
+
+impl Bucketing for Functional {
+    fn sample_to_bucket_minimum(&self, sample: u64) -> u64 {
+        if sample == 0 {
+            return 0;
+        }
+
+        let index = self.sample_to_bucket_index(sample);
+        self.bucket_index_to_bucket_minimum(index)
+    }
+}
+
+impl Histogram<Functional> {
+    /// Create a histogram with functional buckets.
+    pub fn functional(log_base: f64, buckets_per_magnitutde: f64) -> Histogram<Functional> {
+        Histogram {
+            values: HashMap::new(),
+            count: 0,
+            sum: 0,
+            bucketing: Functional::new(log_base, buckets_per_magnitutde),
+        }
+    }
+
+    /// Get a snapshot of _all_ values.
+    pub fn snapshot_values(&self) -> HashMap<u64, u64> {
+        if self.values.is_empty() {
+            return HashMap::new();
+        }
+
+        let mut min_key = None;
+        let mut max_key = None;
+
+        for key in self.values.keys() {
+            let key = *key;
+
+            if min_key.is_none() || key < min_key.unwrap() {
+                min_key = Some(key);
+            }
+
+            if max_key.is_none() || key > max_key.unwrap() {
+                max_key = Some(key);
+            }
+        }
```

Comment:
Could we use [max](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.max)?

Reply:
We could. But we would also need to use `min`. Meaning we run through the list twice.
By doing the iteration ourselves, we can save one trip through the list (I should add a comment).

Reply:
Ah good point.  (/me kinda wants to file a bug to add min_max to Rust stdlib...)